tower-http/src/trace.rs

```diff
@@ -0,0 +1,348 @@
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future, Classifier, EventEmitter, ReqBody>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        let start = Instant::now();
+
+        let span = self.event_emitter.make_span(&req);
+
+        let future = {
+            let _guard = span.enter();
+            self.event_emitter.on_request_received(&req);
+            self.inner.call(req)
+        };
+
+        ResponseFuture {
+            future,
+            span,
+            start,
+            latency_unit: self.latency_unit,
+            classifier: self.classifier.clone(),
+            event_emitter: self.event_emitter.clone(),
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// The [`Future`] produced by [`Trace`] services.
+///
+/// [`Future`]: std::future::Future
+#[pin_project]
+#[derive(Debug)]
+pub struct ResponseFuture<F, Classifier, EventEmitter, ReqBody> {
+    #[pin]
+    future: F,
+    span: Span,
+    start: Instant,
+    latency_unit: LatencyUnit,
+    classifier: Classifier,
+    event_emitter: EventEmitter,
+    _marker: PhantomData<fn() -> ReqBody>,
+}
+
+impl<F, ReqBody, ResBody, E, Classifier, EventEmitter> Future
+    for ResponseFuture<F, Classifier, EventEmitter, ReqBody>
+where
+    F: Future<Output = Result<Response<ResBody>, E>>,
+    Classifier: ClassifyResponse<ResBody, E>,
+    EventEmitter:
+        EmitTracingEvents<ReqBody, OkClass = Classifier::OkClass, ErrClass = Classifier::ErrClass>,
+{
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+
+        let _guard = this.span.enter();
+
+        let result = ready!(this.future.poll(cx));
+        let latency = this.start.elapsed();
+
+        let classification = this.classifier.classify_request_result(&result);
+        this.event_emitter
+            .on_classified_result(classification, latency, *this.latency_unit);
+
+        Poll::Ready(result)
+    }
+}
+
+/// Trait for emitting events and creating spans.
+///
+/// Designed to work with some implementation of [`ClassifyResponse`].
+pub trait EmitTracingEvents<ReqBody> {
+    /// The type used to classify successful responses.
+    ///
+    /// This mirrors [`ClassifyResponse::OkClass`].
+    type OkClass;
+
+    /// The type used to classify failed responses.
+    ///
+    /// This mirrors [`ClassifyResponse::ErrClass`].
+    type ErrClass;
+
+    /// Make a new [`tracing::Span`] from a request.
+    ///
+    /// [`tracing::Span`]: https://docs.rs/tracing/latest/tracing/struct.Span.html
+    fn make_span(&self, req: &Request<ReqBody>) -> Span;
+
+    /// Emit an event when a new request has been received.
+    fn on_request_received(&self, req: &Request<ReqBody>);
```

Comment:
This doesn't actually care about the request body type, does it? I think we don't _need_ the trait to be generic over body types, and could just make the individual methods generic. That might simplify the usage of the trait a bit:
```suggestion
pub trait EmitTracingEvents {
    /// The type used to classify successful responses.
    ///
    /// This mirrors [`ClassifyResponse::OkClass`].
    type OkClass;

    /// The type used to classify failed responses.
    ///
    /// This mirrors [`ClassifyResponse::ErrClass`].
    type ErrClass;

    /// Make a new [`tracing::Span`] from a request.
    ///
    /// [`tracing::Span`]: https://docs.rs/tracing/latest/tracing/struct.Span.html
    fn make_span<ReqBody>(&self, req: &Request<ReqBody>) -> Span;

    /// Emit an event when a new request has been received.
    fn on_request_received<ReqBody>(&self, req: &Request<ReqBody>);
```

Reply:
That was such that if users wanted to do something interesting with the body they could implement this trait for whatever body they have. Could be someone `POST`ing JSON that you want included in your logs. Not sure you can do that when `ReqBody` is declared on the function.

However having the type on the trait does lead to the generic param spreading to a bunch of places which is kinda annoying.

Reply:
it *is* just one generic parameter so it isn't a huge amount of complexity. the future proofing might be worth it.